src/duckling/project/custom-attributes-depot.component.ts

```diff
@@ -0,0 +1,130 @@
+import {Component, ViewContainerRef}  from '@angular/core';
+import {Observable} from 'rxjs';
+
+import {openDialog} from '../util/md-dialog';
+import {ProjectService} from './project.service';
+import {CustomAttribute} from './custom-attribute';
+import {JsonSchema, schemaToJson} from './json-schema.component';
+
+
+/**
+ * Component where the user can add, delete, and edit custom attributes
+ */
+@Component({
+    selector: "dk-custom-attributes-depot",
+    styleUrls: ["./duckling/layout.css", "./duckling/project/custom-attributes-depot.component.css"],
+    template: `
+        <div class="container">
+            <dk-section 
+                class="list-section"
+                headerText="Custom Attributes">
+                <div class="button-container">
+                    <button
+                        *ngIf="curAddingName === undefined"
+                        md-raised-button
+                        title="Add new attribute"
+                        (click)="onNewAttributeAdded()">
+                        Add New Attribute
+                    </button>
+                </div>
+                <div *ngIf="curAddingName !== undefined">
+                    <dk-input
+                        class="dk-inline"
+                        [value]="curAddingName"
+                        [dividerColor]="isValidAttributeName(curAddingName) ? 'primary' : 'warn'"
+                        (inputChanged)="onAttributeNameInput($event)">
+                    </dk-input>
+                </div>
+                <md-list>
+                    <md-list-item *ngFor="let attribute of customAttributes">
+                        {{attribute.key}}
+                    </md-list-item>
+                </md-list>
+            </dk-section>
+            <dk-section 
+                class="adding-section"
+                [headerText]="'Currently adding: ' + addingSectionHeader">
+                <div class="adding-section-body">
+                    <div *ngIf="curAddingSchema !== undefined">
```

Comment:
I don't think you should use curAddingSchema !== undefined as a state test. You should manage the state more explicitly.